include/metal/list/min_element.hpp

```diff
@@ -0,0 +1,59 @@
+#ifndef METAL_LIST_MIN_ELEMENT_HPP
+#define METAL_LIST_MIN_ELEMENT_HPP
+
+#include "../config.hpp"
+#include "../detail/sfinae.hpp"
+#include "../lambda/lambda.hpp"
+#include "../list/list.hpp"
+#include "../number/if.hpp"
+
+namespace metal {
+    namespace detail {
+
+        template<bool done = false>
+        struct _min_element_worker;
+
+        template<>
+        struct _min_element_worker<false> {
+            template<
+                template<class...> class comp, class v0, class v1,
+                class... tail>
+            using type = typename _min_element_worker<!sizeof...(
+                tail)>::template type<comp, if_<comp<v1, v0>, v1, v0>, tail...>;
```

Comment:
Yes, I thought about that. In the ideal case, we find a SFINAE-friendly version of [`combiner`](https://github.com/ecrypa/metal/commit/11289ce3d1a3fad72c5de78c14fce40c1c61b65f#r35182663) and feed that into `fold_left` directly.

If we can not find a SFINAE-friendly `combiner` then the fast-tracking is tedious: Both arguments are mentioned twice (`if_<comp<v1, v0>, v1, v0>`). Do you have an idea how to solve that elegantly?

Reply:
I [found a way](https://github.com/ecrypa/metal/commit/40b2acc65728ba6496723d81f92182c6c14d508b#diff-9689f9ad1946b22fde5ca4180412fdf4), but it is [slow on GCC9](https://ecrypa.github.io/metal-benchmark/min_element_gcc9_update/). I suspect that the aliases are recursively expanded twice for both arguments.

Reply:
> n the ideal case, we find a SFINAE-friendly version of combiner and feed that into fold_left directly.

A SFINAE-friendly version of `combiner` is called [`metal::bind`](https://github.com/brunocodutra/metal/blob/master/include/metal/lambda/bind.hpp#L74). If there were a better way to implement it, then we could just define `min_element` as the beautiful one-liner.

> If we can not find a SFINAE-friendly combiner then the fast-tracking is tedious.

Specially so, because it's an almost exact copy of `metal::fold_left`. Do you think we could extract the internal implementation and make the `c` alias as you defined a parameter such that one could implement both fold_left and min_element (and perhaps other algorithms) without any code duplication?

> it is slow on GCC9.

[There's a trick](https://github.com/ecrypa/metal/commit/40b2acc65728ba6496723d81f92182c6c14d508b#r35379311) you could try.

Reply:
> > it is slow on GCC9.
> 
> [There's a trick](https://github.com/ecrypa/metal/commit/40b2acc65728ba6496723d81f92182c6c14d508b#r35379311) you could try.

Yeah, good idea. Let me call that [v9.foldlike_tempsave](https://github.com/brunocodutra/metal/commit/70c6b77a89cc44083aded0cb503cdd21c53b4cff#diff-d1f453250457743b8a9077c2d785953f). It [does help](https://ecrypa.github.io/metal-benchmark/min_element_gcc9_v4_v10/), but it is still slower than the one-liner. Weirdly, an even faster workaround is given by [v10.foldlike_indirection](https://github.com/brunocodutra/metal/commit/70c6b77a89cc44083aded0cb503cdd21c53b4cff#diff-0a58f9c1c8bba033bd8d808e482e96e9). It is [unbelievably fast on GCC 9](https://ecrypa.github.io/metal-benchmark/min_element_gcc9_v4_v10/), and [it is one of the three equally fast winners on Clang 8](https://ecrypa.github.io/metal-benchmark/min_element_clang8_v4_v10/).

Reply:
> Do you think we could extract the internal implementation and make the `c` alias as you defined a parameter such that one could implement both fold_left and min_element (and perhaps other algorithms) without any code duplication?

Currently, the fastest `min_element` implementation on [Clang 8](https://ecrypa.github.io/metal-benchmark/min_element_clang8_v4_v10/) and [GCC 9](https://ecrypa.github.io/metal-benchmark/min_element_gcc9_v4_v10/) is [v10.foldlike_indirection](https://github.com/brunocodutra/metal/commit/70c6b77a89cc44083aded0cb503cdd21c53b4cff#diff-0a58f9c1c8bba033bd8d808e482e96e9). If I remember correctly then you should be able to change the combiner (`c`) from
```
template<template<typename...> class comp, class x, class y>
using _min_element_combiner = if_<comp<y, x>, y, x>;
```
to
```
template<template<typename...> class expr, class x, class y>
using _fold_left_combiner = expr<x, y>;
```
in order to obtain an implementation of `metal::fold_left`. I suspect that it should be equally fast, but I hesitate to touch such a fundamental well tested algorithm.

Reply:
> It is unbelievably fast on GCC 9, and it is one of the three equally fast winners on Clang 8.

Wow amazing!

> If I remember correctly then you should be able to change the combiner (c) from

You're definitely onto something. It's possible this indirection would trip MSVC, but worst case we can implement something analogous to `detail::forward`.

Reply:
> If I remember correctly then you should be able to change the combiner [...]

It seems to work. See #100. I am not sure if it is a good idea, though.